irohad/ordering/impl/on_demand_ordering_service_impl.cpp

```diff
@@ -15,13 +16,53 @@
 #include "datetime/time.hpp"
 #include "interfaces/iroha_internal/proposal.hpp"
 #include "interfaces/iroha_internal/transaction_batch.hpp"
+#include "interfaces/iroha_internal/transaction_batch_impl.hpp"
+#include "interfaces/iroha_internal/transaction_batch_parser_impl.hpp"
 #include "interfaces/transaction.hpp"
 #include "logger/logger.hpp"
 #include "main/subscription.hpp"
+#include "ordering/ordering_types.hpp"
 #include "subscription/scheduler_impl.hpp"
 
 using iroha::ordering::OnDemandOrderingServiceImpl;
 
+namespace {
+  auto parseProposal(
+      shared_model::interface::types::TransactionsCollectionType const &txs) {
+    shared_model::interface::types::SharedTxsCollectionType transactions;
+    for (auto const &transaction : txs)
+      transactions.push_back(clone(transaction));
+
+    return shared_model::interface::TransactionBatchParserImpl().parseBatches(
+        transactions);
+  }
+
+  void uploadBatches(
+      iroha::ordering::BatchesCache::BatchesSetType &batches,
+      shared_model::interface::types::TransactionsCollectionType const &txs) {
+    auto batch_txs = parseProposal(txs);
+    for (auto &txs : batch_txs) {
+      batches.insert(
+          std::make_shared<shared_model::interface::TransactionBatchImpl>(
+              std::move(txs)));
+    }
+  }
+
+  void uploadBatchesWithFilter(
+      iroha::ordering::BloomFilter256 const &bf,
+      iroha::ordering::BatchesCache::BatchesSetType &batches,
+      shared_model::interface::types::TransactionsCollectionType const &txs) {
+    auto batch_txs = parseProposal(txs);
+    for (auto &txs : batch_txs) {
+      auto batch =
+          std::make_shared<shared_model::interface::TransactionBatchImpl>(
+              std::move(txs));
+      if (bf.test(batch->reducedHash()))
```

Comment:
Is it possible to have false-positive in bloom filter? How to handle the situation?